### ISSUE TITLE
ttnn.argmax: FLOAT32 never selects a NaN and BFLOAT16 lets a later +0 beat an earlier -0, so the two dtypes return different indices and both differ from torch

### ISSUE THREAD

**Issue by ping-dev-ui** [tenstorrent/tt-metal#55346]
**Component / Area:** ttnn reduction, `argmax` reader kernels and the shared float comparators

**Issue Type:** Bad Outputs

## Summary

The non-SFPU argmax paths (single core, multicore, tile-layout) compare raw bit patterns through two helpers:

- [float32.h#L18-L61](https://github.com/tenstorrent/tt-metal/blob/9fa00fce13cda1538f0130695576408ff5b05665/tt_metal/hw/inc/api/numeric/float32.h#L18-L61)
  `float32_greater`: "NaN is never greater than anything" (returns false if either operand is NaN); `±0` compare equal.
- [bfloat16.h#L66-L92](https://github.com/tenstorrent/tt-metal/blob/9fa00fce13cda1538f0130695576408ff5b05665/tt_metal/hw/inc/api/numeric/bfloat16.h#L66-L92)
  `bfloat16_greater`: no NaN and no zero handling; sign-first ordering, so `+0 (0x0000)` is strictly greater than `-0 (0x8000)`.

Consumers: [argmax_common.hpp#L130-L137](https://github.com/tenstorrent/tt-metal/blob/9fa00fce13cda1538f0130695576408ff5b05665/ttnn/cpp/ttnn/operations/reduction/argmax/device/kernels/argmax_common.hpp#L130-L137)
(`update_max_if_greater`, strict greater only) and `#L218-L232` (`process_core_data`, tie test is a raw `val == max_val`),
used by `reader_argmax_interleaved.cpp`, `reader_argmax_interleaved_multicore.cpp`, `argmax_tile_h_col.hpp` and
`argmax_tile_layout.hpp`.

The registered golden is `torch.argmax`
([reduction.py#L801](https://github.com/tenstorrent/tt-metal/blob/9fa00fce13cda1538f0130695576408ff5b05665/ttnn/ttnn/operations/reduction.py#L801)),
which treats NaN as the maximum (consistent with `torch.max` propagating NaN) and returns the first index among equal
maxima, with `-0.0 == +0.0`.

This is a source-level analysis; the comparators are plain integer C++ and the replay is a line-for-line port, so the
device result is the replay result.

## Observed vs expected

| input row | torch (f32 and bf16) | ttnn FLOAT32 | ttnn BFLOAT16 |
|---|---|---|---|
| `[3, nan]` | 1 | **0** | 1 |
| `[nan, 3]` | 0 | **1** | 0 |
| `[1, nan, 3]` | 1 | **2** | 1 |
| `[-inf, nan]` | 1 | **0** | 1 |
| `[-0.0, 0.0]` | 0 | 0 | **1** |
| `[-0.0, 0.0, -1]` | 0 | 0 | **1** |
| `[0.0, -0.0]` | 0 | 0 | 0 |
| `[2, 2]` | 0 | 0 | 0 |

The same row gives a different index in FLOAT32 than in BFLOAT16. The BFLOAT16 "NaN wins" behaviour is accidental: the
positive quiet NaN `0x7FC0` just has a larger bit pattern than `+inf`; a negative-signed NaN (`0xFFC0`) orders below
`-inf` and is never selected.

## Why the tests do not catch it

`tests/ttnn/unit_tests/operations/reduce/test_argmax.py` draws inputs from `torch.rand`, which never produces NaN or
`-0.0`. Related but different: #53769 (multicore work split), #21120 (`keepdim=False`), #48188 item 8 (`reduce_core_id`
bool cast). None mention NaN or signed zero.

## Reproducer (host only, `pip install torch`)

```python
import struct, torch
NAN = float("nan")
F32_SIGN, F32_EXP, F32_MAN, F32_MAG, BF_SIGN = 0x80000000, 0x7F800000, 0x007FFFFF, 0x7FFFFFFF, 0x8000

def float32_greater(a, b):                                  # float32.h L18-61
    if ((a & F32_EXP) == F32_EXP and a & F32_MAN) or ((b & F32_EXP) == F32_EXP and b & F32_MAN):
        return False
    if (a & F32_MAG) == 0 and (b & F32_MAG) == 0:
        return False
    if (a ^ b) & F32_SIGN:
        return (a & F32_SIGN) == 0
    return a < b if (a & F32_SIGN) else a > b

def bfloat16_greater(a, b):                                 # bfloat16.h L66-92
    if (a ^ b) & BF_SIGN:
        return (a & BF_SIGN) == 0
    return a < b if (a & BF_SIGN) else a > b

f32 = lambda x: struct.unpack("<I", struct.pack("<f", x))[0]
bf16 = lambda x: torch.tensor(x, dtype=torch.bfloat16).view(torch.int16).item() & 0xFFFF

def ttnn_argmax(vals, bits, greater, neg_inf):              # update_max_if_greater over the row
    idx, mx = 0, neg_inf
    for i, v in enumerate(map(bits, vals)):
        if greater(v, mx):
            idx, mx = i, v
    return idx

for vals in ([3., NAN], [NAN, 3.], [1., NAN, 3.], [-0.0, 0.0]):
    print(vals, "torch", torch.argmax(torch.tensor(vals)).item(),
          "ttnn f32", ttnn_argmax(vals, f32, float32_greater, 0xFF800000),
          "ttnn bf16", ttnn_argmax(vals, bf16, bfloat16_greater, 0xFF80))
```

## Proposed fix

Give both comparators torch semantics: `is_nan(a) && !is_nan(b) -> true`, `is_nan(b) -> false`, and treat two zeros of
either sign as equal (already done in `float32_greater`, missing in `bfloat16_greater`). In `process_core_data` and
`process_value_comparison`, replace the raw `val == max_val` tie test with a value-equality helper (`±0` equal, NaN equal
to NaN) so the first-index rule holds. Add `[-0.0, 0.0]` and NaN rows to `test_argmax.py` for both dtypes and for the
N/C-dim SFPU path (whose `gt_binary_tile` NaN behaviour I could not check on the host and should be covered by the same
test).

## Acceptance criteria (proposed)

- `ttnn.argmax` matches `torch.argmax` on rows containing NaN, `-0.0`, `+0.0` and ties, for FLOAT32 and BFLOAT16, on the
  single-core, multicore and tile-layout paths.
- FLOAT32 and BFLOAT16 return the same index for the same data.
- `test_argmax.py` extended with those rows.

**Comment by ping-dev-ui** [comment]
Reporter here. Fix with tests is ready on my fork (https://github.com/ping-dev-ui/tt-metal/tree/fix/argmax-nan-signed-zero): both bit-pattern comparators now follow torch (NaN is the maximum, +0 == -0), new bfloat16_equal / float32_equal helpers back the tie tests so the first index wins among equal maxima, and the tests cover NaN, signed-zero and tie rows on the width path (ROW_MAJOR and TILE), the tile-layout height path and the multi-core reduce_all path in bf16 and fp32. Holding the PR until assignment per the bounty terms. I have read and agree to the Tenstorrent Bounty Program Terms & Conditions (https://docs.tenstorrent.com/bounty_terms.html). @jberkowitzTT could you assign this to me when it is reviewed?

**Comment by heidman06** [comment]
Implemented in PR #55350 with PyTorch NaN and signed-zero comparator semantics for float32 and bfloat16, values_equal tie helper, and unit tests covering NaN and [-0.0, 0.0] rows.

**Comment by jberkowitzTT** [comment]
Hi @ping-dev-ui - please submit all bounty proposals to [bounties@tenstorrent.com](mailto:bounties@tenstorrent.com). If approved, we'll create the issue and assign it to you as a bounty

Thank you for your enthusiasm & understanding

**Comment by heidman06** [comment]
Thanks @jberkowitzTT, submitted the proposal via email as well. Much appreciated!
